### PR TITLE
Use single balances in the Stellar Asset Contract

### PR DESCRIPTION
**Pull request by sisuresh**
### Changes

This PR updates how balances are stored for **Stellar accounts**. Stellar account balances must be stored in classic trustlines, and classic trustlines semantics are followed. For example-

1. Transfers involving a Stellar account will only succeed if the trustline has the authorized flag set.
2. Burn (will be changed to clawback in my next PR) can only be called if the clawback flag is set on the trustline.
3. The admin can only call freeze (will be changed to deauthorize in my next PR) on a Stellar account if the issuer has auth_revocable set.
4. Issuers don't have trustlines, so if the issuer sends a balance, that balance will be minted. If the issuer receives a balance, that balance will burned. We should consider how downstream system will handle this case, and maybe emit mint/burn events instead.

This PR also makes a minor improvement related to burn(clawback) - it can be called by the admin even if the balance/trustline is frozen/deauthorized.

### Improvements I'll work on while this PR is being reviewed
1. Consider improving error reporting when trustlines are missing, as I expect this case will be hit much more often now.
2. Continue writing tests.
3. Consider keying non-account balances with an `Identifier`-like struct that doesn't include `Account` for safety.

### Next PR will update the interface
The follow up to this will be a changes that updates to interface to make the terminology match more closely with classic, and will resolve some inconsistencies in this PR like  -
1. burn/clawback
2. freeze/deauthorize
3. read_state/write_state (these are internal methods)

### CODE REVIEW

File: soroban-env-host/src/native_contract/token/balance.rs [hunk 1]
```diff
@@ -5,20 +5,33 @@ use crate::native_contract::token::public_types::{Identifier, Metadata};
...
 use super::error::ContractError;
 
 // Metering: *mostly* covered by components. Not sure about `try_into_val`.
 pub fn read_balance(e: &Host, id: Identifier) -> Result<i128, HostError> {
-    let key = DataKey::Balance(id);
-    if let Ok(balance) = e.get_contract_data(key.try_into_val(e)?) {
-        Ok(balance.try_into_val(e)?)
-    } else {
-        Ok(0)
+    match id {
```
Comment (dmkozh) [inline soroban-env-host/src/native_contract/token/balance.rs:17]: Just to double-check: have we agreed that we should expose the full balance by default and not vice versa? My intuition is that spendable balance will have more users (contracts), which is why it could be better to make it default.

Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: I wouldn't say `balance` is the default. It's just an accurate name for the `Identifiers` current balance. Contracts would just used `sp_balance` instead if they want the spendable balance (I hate this name but 10 char limit :cry:. Hopefully we do something like [this](https://github.com/stellar/rs-soroban-env/issues/584)).

I think we also discussed this in a protocol meeting and the consensus was that we should add a spendable balance, but I don't remember who mentioned that.

File: soroban-env-host/src/native_contract/token/balance.rs [hunk 2]
```diff
@@ -5,20 +5,33 @@ use crate::native_contract::token::public_types::{Identifier, Metadata};
 use crate::native_contract::token::storage_types::DataKey;
 use crate::{err, HostError};
 use soroban_env_common::xdr::{
-    AccountEntryExt, AccountEntryExtensionV1Ext, AccountId, LedgerEntryData, TrustLineAsset,
-    TrustLineEntryExt, TrustLineFlags,
+    AccountEntry, AccountEntryExt, AccountEntryExtensionV1Ext, AccountFlags, AccountId,
```
Comment (sisuresh) [inline soroban-env-host/src/native_contract/token/balance.rs:8]: I think this file should be refactored to separate trustline vs non-trustline logic. We can do that in a separate PR.

File: soroban-env-host/src/test/token.rs [hunk 3]
```diff
@@ -2208,7 +2339,7 @@ fn test_asset_token_classic_balance_boundaries_with_limit_and_liabilities() {
 #[test]
 fn test_asset_token_classic_balance_boundaries_large_values() {
     test_wrapped_asset_classic_balance_boundaries(
-        i64::MAX,
+        i64::MAX - i64::MAX / 5,
```
Comment (dmkozh) [inline soroban-env-host/src/test/token.rs:2342]: Why did this change?

File: soroban-env-host/src/native_contract/token/balance.rs (hunk 1 above)
Comment (leighmcculloch) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: You could alternatively use the name `spendable`.

I think we've discussed this a few times, and each time someoen has made a compelling argument that the default should be the full balance. I don't remember the full reasons of why we've said that before. But for me it occurs this way because:
- It is more important for the user that the default is their full so that the user doesn't think they've lost money when it is not spendable.
- It is consistent with how financial institutions communicate about balances. They typically use the terms `balance` and `available balance`.

Maybe we should use the term `available` instead of `spendable`.

Comment (leighmcculloch) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: > Also it's a bit annoying that the standard token interface would need to have two balance methods

I think we'll find custom contracts will have reasons to use it, and it is a small inconvenience to code two very lightweight methods. I think it is okay.

Comment (dmkozh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: > It is more important for the user that the default is their full so that the user doesn't think they've lost money when it is not spendable.

Who is the 'user' here? If it's the wallet, it may be using e.g. an RPC endpoint. I don't think that a case of calling a balance fn on the contract would be frequent enough.

My intuition is that the main users of the token contract interface are other contracts (as in 'unique users', i.e. the number of custom contracts is way bigger than the number of the custom wallets and those will probably use an even smaller number of services to get balance). I'm reasonably certain that a large percentage of the contract writers would just expect that they should use `balance`, as in ERC-20 tokens.

File: soroban-env-host/src/native_contract/token/balance.rs [hunk 4]
```diff
@@ -31,77 +44,216 @@ fn write_balance(e: &Host, id: Identifier, amount: i128) -> Result<(), HostError
...
     Ok(())
 }
 
+// TODO: Metering analysis
+//returns (total balance, spendable balance)
+pub fn get_classic_balance(e: &Host, to_key: AccountId) -> Result<(i64, i64), HostError> {
+    match read_metadata(e)? {
+        Metadata::Native => get_account_balance(e, to_key),
+        Metadata::AlphaNum4(asset) => {
+            if asset.issuer == to_key {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:233]: I'm actually not clear on what this helper would look like, and how it would be used here. The asset type is different between the match cases and we're just checking the issuer. Are you saying we should have a helper that returns the `TrustLineAsset` and issuer based on the metadata?

File: soroban-env-host/src/test/token.rs (hunk 3 above)
Comment (sisuresh) [reply soroban-env-host/src/test/token.rs:2342]: This test specifies `i64::MAX / 5` amount of buying liabilities, which means the max balance should be `i64::MAX - i64::MAX / 5`.

Comment (sisuresh) [reply soroban-env-host/src/test/token.rs:2342]: I added a step in `test_wrapped_asset_classic_balance_boundaries` that sends the balance back and it fails there.

File: soroban-env-host/src/native_contract/token/balance.rs [hunk 5]
```diff
@@ -31,77 +44,216 @@ fn write_balance(e: &Host, id: Identifier, amount: i128) -> Result<(), HostError
...
-            .checked_add(amount)
-            .ok_or_else(|| e.err_status(ContractError::OverflowError))?;
-        write_balance(e, id, new_balance)
+        return Err(e.err_status_msg(ContractError::BalanceFrozenError, "balance is frozen"));
+    }
+
+    match id {
+        Identifier::Account(acc_id) => {
+            let i64_amount = i64::try_from(amount)
+                .map_err(|_| e.err_status_msg(ContractError::OverflowError, "greater than i64"))?;
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs]: Updated

File: soroban-env-host/src/native_contract/token/balance.rs (hunk 1 above)
Comment (dmkozh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: Another angle would be that the contracts shouldn't worry about the spendable balance at all and everyone should just use the full balance. It's up to the token to decide whether the token can be spent, so the classic-specific stuff is just one of the ways the transfer may fail, but there are also others.

File: soroban-env-host/src/native_contract/token/balance.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:241]: Oh wait I see what you're saying. I should use a method that takes the issuer and TrustLineAsset so all special logic is shared.

[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/native_contract/token/balance.rs:241]: Yeah, exactly, It wouldn't save too many lines, but would make maintenance/review way easier.

File: soroban-env-host/src/native_contract/token/balance.rs (hunk 1 above)
Comment (leighmcculloch) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: I think that is a very reasonable approach too, and I think it encourages a certain level of decoupling of concerns. We can add an available balance later on as an extension to the interface if it turns out we were wrong and we really needed it.

File: soroban-env-host/src/native_contract/token/balance.rs [hunk 6]
```diff
@@ -31,77 +44,216 @@ fn write_balance(e: &Host, id: Identifier, amount: i128) -> Result<(), HostError
...
+    match id {
+        Identifier::Account(acc_id) => match read_metadata(e)? {
+            Metadata::Native => {
+                return Err(e.err_status_msg(
+                    ContractError::OperationNotSupportedError,
+                    "cannot clawback native asset",
+                ))
+            }
+            Metadata::AlphaNum4(asset) => {
+                if asset.issuer == acc_id {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs]: Done

File: soroban-env-host/src/native_contract/token/balance.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs]: Done

File: soroban-env-host/src/native_contract/token/balance.rs (hunk 1 above)
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: I don't think there's an issue with expecting everyone to use the total balance for now, but is there an issue with providing both? Without a spendable balance, a contract doing `xfer_from` wouldn't be able to use `min(balance, allowance)` to at least pull a partial amount. Maybe this is an edge case, but it sounds like a reasonable thing to do.

I don't have an issue with removing spendable balance for now, but I'm curious if contracts will suffer without it.

Comment (MonsieurNicolas) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: yeah spendable balance is what contracts want to use, the other one is more like a UX thing for wallets

Comment (MonsieurNicolas) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: We should rename this one to lift the ambiguity (people are likely to use it without reading docs if it's just called "Balance"), something like "TotalBalance" or something like that.

Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: The 10 character limits us from doing `TotalBalance`. We could do `total` and `spendable`, but not having `balance` anywhere might confuse users.

Comment (dmkozh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: IMHO more similar methods = more confusion, given that the difference is very subtle. Your example is interesting, but I'm not sure if that's something the contract do in reality instead of just doing what's requested. I guess it's a bit of a philosophical question. Imagine you deposit (A, B) amounts of tokens, but only (0.5A, 0.5B) gets deposited; is that what you expect from the call? I would expect most financial operations to be pretty strict/deterministic and failing feels safer most of the time if something is off.
Either way, I guess I'm convinced that it's ok to return the full balance for `balance`. I'd probably vote for removing the spendable balance for now to keep interface more slim/clear.

Comment (dmkozh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: Also, to come back to the example: if we go with account abstraction approach, then the wallet may also have control over the amount of token spent. It *probably* would happen before the contract logic kicks in, but in general that's yet another indication that the best strategy for the contract would be to call into high level interface methods and not do any assumptions on what should make them successful (unless that's explicitly a part of the interface). Also the token implementation itself may impose the limits on the transfers. Basically I don't think that there are any requirements to the token interface that would tie balance and xfers.

Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: My example is a little different than yours. If a specific amount was requested, than that amount has to be used. I was think of scenarios where you just specify an account and pull what you can. Again, not sure if this type of use case will be common. Also, I don't think it's unreasonable to specify in the interface that a `xfer` with spendable balance should never fail due to insufficient funds.

But anyways, in the interest of closing on this, I think we should just leave both methods for now, but change the names to something like `total_bal` and `spndbl_bal`. When we eventually allow longer function names, we can come back and rename. We can also revisit both methods after we see how these contracts will be used. Thoughts on this and the naming?

Comment (dmkozh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: But the user still needs to provide the max amount, don't they? It's probably not a good idea to give the infinite allowance. So in this scenario available balance seems to actually be more useful off-chain to figure out the proper allowance... Anyway, I don't feel that strongly about this; if you think it's worth it, we can keep both for the time being. The only thing to keep in mind is that it's much easier to extend the interface and probably pretty much impossible to shrink it.

Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: The contract doing the `xfer_from` can call `allowance` to get whatever the user has approved.

When you say impossible to shrink it, you're referring to after we officially launch right? We can still change the interface over the next couple months.

Comment (leighmcculloch) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: Good point. Let's just keep both. This is a small detail. We can change it later if we're wrong.

Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs:17]: Updated `sp_balance` to `spendable` for now.

**Review by dmkozh — APPROVED** [review]

File: soroban-env-host/src/native_contract/token/balance.rs [hunk 7]
```diff
@@ -5,20 +5,33 @@ use crate::native_contract::token::public_types::{Identifier, Metadata};
...
+            }
+        }
+    }
+}
+
+// Metering: *mostly* covered by components.
+pub fn get_spendable_balance(e: &Host, id: Identifier) -> Result<i128, HostError> {
+    match id {
+        Identifier::Account(acc_id) => Ok(get_classic_balance(e, acc_id)?.1.into()),
+        _ => read_balance(e, id),
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-env-host/src/native_contract/token/balance.rs]: Good point. Done.
